### ISSUE TITLE
Display and Edit Asset citations

### ISSUE THREAD

**Issue by NuclearRedeye** [libero/editor#387]
This issue covers the implementation of support for the `<xref ref-type="fig"> <xref ref-type="supplementary-material"> <xref ref-type="video"> <xref ref-type="table"> ` elements.

# User Stories

https://github.com/libero/editor/issues/45

### Additional User Stories
- [ ] As an **author**, I want to be able to add a citation to a figure, table or another asset so that I can reference a relevant item.
- [ ] As an **author**, I want to be able to remove an asset citation so that I can correct mistakes.
- [ ] As an **author**, I want to be able to edit an existing citation so that I can correct errors.

# Design
- [Asset citation prototype](https://invis.io/A8WXRN9PQBF#/420310643_figure-Citation)

# DoD
- [ ] Editor updated so that the user can see and edit all the required fields.
- [ ] Solution **reviewed** and **approved** by the designer.
- [ ] Solution **reviewed** and **approved** by required stakeholders.
- [ ] New version of the Editor, deployed to the staging environment and publicly accessible.

**Comment by NuclearRedeye** [comment]
### Estimate
Design: 1d
Dev: 2d

**Comment by fred-atherden** [comment]
I'm seeing some pretty weird behaviour when I highlight the a figure citation (see [attached](https://app.zenhub.com/files/195217312/49596b4d-249b-4e5c-b5e2-39c61aab9414/download)).

**Comment by fred-atherden** [comment]
As discussed on the call - we need the ability to add a citation without highlighting text first.

Ideally this could work in the same way as reference citations (`???` is added, and then once you select the Figure you want to cite, that Figure label is inserted as the text for the citation).

**Comment by fred-atherden** [comment]
I'm able to add a figure citation that doesn't link to any figure, which is problematic.

**Comment by fred-atherden** [comment]
Deleting a figure should remove all the figure citations that link to that figure.

I think we'll also need a warning pop-up for any Figure that has figure citations similar to what we currently have for references:
![Screenshot 2020-11-10 at 10.43.26.png](https://images.zenhubusercontent.com/5d5a77bfe56ed603cbb191f6/cf320f90-e400-4626-8b74-fd3c25209777)

**Comment by NickDuf** [comment]
@FAtherden-eLife @JGilbert-eLife @ilya-zhuravel Some unusual UX at present, and I probably need to work through these ideas with production, before we can address anything.

- If I don't name a figure we get an option from the figure citation menu without text
- Citing a figure is inconsistent with citing a reference which can cause confusion (without highlighting text there is no option available. As Fred has mention the little ??? will fix this issue)
- If I swap the figure labels on the page we then have figure 2 before figure 1 in the list. I understand why we have done this for the time being, but the UX feels broken and will make scrolling the list difficult for users looking for figures 

![Screenshot 2020-11-10 at 10.42.11.png](https://images.zenhubusercontent.com/5d5a52c42298e3000103df57/0c10a392-a8b3-4f33-a10a-1a2a87808b80)

**Comment by FAtherden-eLife** [comment]
Filter list doesn't work as expected:
![Screenshot 2020-11-10 at 10.56.39.png](https://images.zenhubusercontent.com/5d5a77bfe56ed603cbb191f6/ec76a11a-1698-455a-a46d-b26c3074f4d5)

**Comment by ilya-zhuravel** [comment]
@FAtherden-eLife 
1. I managed to reproduce the dragging issue once. Not very consistent but it seems to be related to browser/prosemirror/other. Let's raise a separate ticket for it.
2. I'll make it work with empty selections, no problem, but it goes against 3.
3. Just like reference a citation is a link to a specific object. If we have reference citations with question marks (not linked to anything) why can't we figure citations that aren't linked to any figure? What would be preferred behavior in your case?
4. I'll make sure to update citations when figure is removed. What happens with citations that link to just the deleted figure? Should the citation be removed (with the text inside it potentially)? Should it just become empty citation? This also refers to #3 in this list. Not allowing empty citations can potentially lead to some complex UX-related issues. Perhaps we should just mark empty citations (for both assets and references) as invalid and highlight the text with red? Let's confirm the pop-up alert with @NickDuf. Probably best to add a new ticket and include title and message for the pop up.

@NickDuf 
5. There're aren't many options to fill the empty box if a figure number is missing. At the moment the figure number comes from the corresponding field and user can leave it empty. An alternative would be to have a placeholder text e.g. 'unnamed figure #371' where #371 is an arbitrary number, and could be the number of that figure as it appears in the main text. Or we could try to scoop up any text that has been entered for that figure including title, legend, attribution and license, but the former case still stays, because any inserted figure will have no content at all and will need to have a kind of placeholder.
6. I'll add figure citations for an empty selection, no worries.
7. Currently figures in citation popup appear in the order they are discovered in the main text. I can't think of any other way of organising that list except sorting it alphabetically, but then if the names are wouldn't it lead to even more confusion? Let me know what you decide.

**Comment by NickDuf** [comment]
@ilya-zhuravel some of the issues may be solved post MVP as we start to look at asset numbering generally. A placeholder is definitely the obvious option for an unlabelled figure. Let's have a think an agree a short-tern solution.

**Comment by fred-atherden** [comment]
Thanks @ilya-zhuravel. 

> I managed to reproduce the dragging issue once. Not very consistent but it seems to be related to browser/prosemirror/other. Let's raise a separate ticket for it.

Sure, I've raised #598. 

> What happens with citations that link to just the deleted figure? Should the citation be removed (with the text inside it potentially)?

Yup, any citation (including the text) to the deleted figure should also be deleted.

> Not allowing empty citations can potentially lead to some complex UX-related issues. Perhaps we should just mark empty citations (for both assets and references) as invalid and highlight the text with red? Let's confirm the pop-up alert with @NickDuf. Probably best to add a new ticket and include title and message for the pop up.

Yeah this is a good point. Maybe we can re-asses once the `???` functionality has been added. I like the idea of indicating to the user that something's wrong (such as with red highlighting or a pop-up). Imagine that's something that we should do as part of the 'validation' epic, which I understand is going to be handled post-MVP. So maybe let's skip over this point for now.

**Comment by fred-atherden** [comment]
When inserting a Figure citation using the `???` functionality, would it be possible for the link text to change from `???` to the label of the figure (without the full stop at the end - if there is one) on click?

For example here:
![Screenshot 2020-11-11 at 11.29.04.png](https://images.zenhubusercontent.com/5d5a77bfe56ed603cbb191f6/3d94711c-7b41-4200-b611-b5a1248820ce)
The link text would change to `Figure 1`

**Comment by NickDuf** [comment]
@ilya-zhuravel remaining to-do's @FAtherden-eLife does this cover everything?

• Add placeholder for unlabelled figure in citation menu 'Unlabelled figure'. If option is chosen placeholder text is used for citation.
• ??? Affordance is updated to item chosen on click
• If cited figures are deleted use same prompt for delete reference, replacing the word 'reference' with 'figure'

**Comment by fred-atherden** [comment]
Thanks @NickDuf, yup that looks good. 
Just to make explicit RE this:
> • ??? Affordance is updated to item chosen on click

The label text without the full stop at the end. So if the figure label is `Figure 1.`, then `Figure 1` is included as the link text.

[Edit text of this comment updated]

**Comment by fred-atherden** [comment]
These two are fine:
> • ??? Affordance is updated to item chosen on click
> • If cited figures are deleted use same prompt for delete reference, replacing the word 'reference' with 'figure'

This one I am unable to test currently:
> • Add placeholder for unlabelled figure in citation menu 'Unlabelled figure'. If option is chosen placeholder text is used for citation.

Because editing any field in a figure/reference disables the insert toolbar functions. Similarly inserting a citation disables the insert toolbar functionality after it (meaning that you cannot insert two or more citations).

[https://app.zenhub.com/files/195217312/82d40e51-380c-4aab-9e4c-d8bebd7113b4/download](https://app.zenhub.com/files/195217312/82d40e51-380c-4aab-9e4c-d8bebd7113b4/download)
